### DISCUSSION TITLE
Just a thought on levels of testing

### DISCUSSION THREAD

**Discussion by Naktakala** [chi-tech/chi-tech#174] [General]
I want to manage testing as follows:
- Unit tests. Yes they are a pain but if you want versatility and robustness they need to be added to code pieces that are expected to serve a broad scope. We are lacking in this. I am still learning how to do them properly. The biggest hurdle at this moment is resources. It is easy to point at codes such as PETSc, hypre, trilinos, deal.ii and go like "hey we should be doing this" but in reality we face the same challenges that PDT faced. Our team's performance is based on the simulations we do and not the API's robustness ... essentially **making unit test writing a resource sink** with a very low weight for our KPI's.

- Regression Tests. Exactly what the word says ... testing whether something that worked is no longer working. The code base regressed. A very important aspect for speed of development. I am largely happy with them for now since they've served their purpose well. They need to be expanded yes, and they need to be queue-able so that you can run a little faster. They should have the following properties:
  1. Not take too long so they become an inconvenience rather than a life-saver
  2. Run-able on a personal computer
  3. Should test the main-stream functionality (not just corner cases) and most used features (not saying they can't test corner cases)
  4. They need to test for correct answers, iteration counts, memory consumption and maybe even timing, not just to check for crashes

- Verification tests. They have the following properties:
  1. Generally more extensive suits of tests verifying that the code is operating correctly and producing correct/acceptable answers. 
  2. They may be real world problems.
  3. They may take a long time to run or require a lot of computing resources.
  4. They should be well documented and usable by code users to build validation cases.

**Comment by ragusa** [comment]
Regression Tests: a must have. I agree with the description.

Verification test: I do not think verification is the proper term, is it? To me verification when coding is verifying that the equations are solved correctly, that is, that we get the proper convergence rate, proper spectral radius, ... Here, what you are describing is what I have seem labeled as "heavy" regression tests. The kind of tests that are run only once a day (nightly) or once a week, to make sure that recent changes didn't break them.

Unit tests: Gosh, I should not be writing this here, on a public discussion forum, but I agree with you. This is a must but we don't have the bandwidth. :-(

**Reply by Naktakala** [reply]
We can add "Heavy regression tests" to this list. But they should be different from verification tests. Let me explain.

Apart from good theory manuals, all the black-box or otherwise legendary software codes have a well documented set of verification suites. MCNP has them, ANSYS-structural analysis has them, STAR-CCM+ has them, etc. They are absolutely golden nuggets to build validation cases for their intended use.

I want to use verification tests in the same setting as international standards uses them. End-user level. I've used STAR-CCM+ in a validated envelope for reactor cooling simulations. It was certified according to ASME-NQA1. What we had to do was to use the well documented verification tests (shipped with the software) to build a validation suite (problems + literature + experimental measurements) which established the applicable envelope in which we could use the code for safety calculations. For the SAFARI-1 reactor this was sub-cooled light-water with heat transfer and no nucleate boiling.

Does this make sense? Basically Everything else is for developers, verification is for the end-user and implies VERY WELL documented.

**Comment by ragusa** [comment]
Ok. There should be tests for developers:

  *   Units tests (but no bandwith now)
  *   Light-weight reg tests (to be passed often, each PR?). Some of those are sanity testing.
  *   Heavy reg tests  (to be passed regularly, nightly? Weekly?)

For users:

  *   Tutorials
  *   And those tests which I didn’t know were called verification tests. They could also be called integration tests as they are comprehensive, right?
In the present PR, an infinite medium k-eigenvalue, in my opinion, is not a verification test. To me, it falls in the category of quick-to-run reg tests.

**Reply by Naktakala** [reply]
Yes. Light weight reg tests should be run-able as often as possible. Heavy reg tests could become a weekly thing.

An infinite medium k-eigenvalue test can be both a reg test and a verification test. I think MCNP has something like it.
